plugins/crate-copy-azure/src/main/java/io/crate/copy/azure/AzureFileInput.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.copy.azure;
+
+import static io.crate.copy.azure.AzureFileOutput.azureResourcePath;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.opendal.Entry;
+import org.apache.opendal.Operator;
+import org.elasticsearch.common.settings.Settings;
+import org.jetbrains.annotations.Nullable;
+
+import io.crate.execution.engine.collect.files.FileInput;
+import io.crate.execution.engine.collect.files.Globs.GlobPredicate;
+import io.crate.execution.engine.collect.files.opendal.Scheme;
+
+/**
+ * File reading components operate with URI.
+ * All URI-s in the public API follow the contract "outgoing/incoming" URI is Azure compatible.
+ * This is accomplished by transforming user provided URI to the Azure compatible format only once.
+ * Outgoing URI-s are then used by other components and sent back to this component,
+ * so outgoing format (expandURI) implicitly dictates incoming URI-s format (getStream).
+ */
+public class AzureFileInput implements FileInput {
```

Comment:
Maybe something for a follow up where we migrate s3 to OpenDAL too, but I was hoping that the per-scheme implementation becomes much smaller by using it.

E.g. We could have a `copy` module which defines a `OpenDALInput` which can be parameterized with the supported scheme/properties and anything that can be specific to a backend. It looks like only the URI building needs some custom logic?

Reply:
> It looks like only the URI building needs some custom logic?

Yes

So basically you mean moving out as much as code as possible to the common module and make OpenDAL dependant plugins work only with non-common logic?

It makes thanks, will do (as a follow if it's fine), want to finish working MVP first as this refactoring won't affect user facing functionality.

Just to be sure:

Given that Azure plugin is the only one using OpenDAL currently, total amount of code won't change.

> per-scheme implementation becomes much smaller 

will be true only once we migrate S3 or/and add new plugins for other providers

Reply:
> (as a follow if it's fine).

yes, probably better anyway to do it in another step - together with migrating s3. Always helps to have 2-3 implementations to validate an abstraction.